### PR TITLE
Implement `describeType(null)` stub

### LINKED ISSUE #10143: Burrito Bison: Stuck before title screen
### Describe the bug

https://www.newgrounds.com/portal/view/562620?emulate=flash
Tested it out on the web page and tested it on desktop via spoofing the web page to get past the site lock, the results still the same, the game gets stuck on the screen here and nothing shows up in the console logs.

Burrito Bison Ruffle

### Expected behavior

The game should actually start

### Affected platform

Browser's extension

### Operating system

Windows 11

### Browser

Edge 110

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Coder-256**
With this patch, Burrito Bison now works all but flawlessly in my testing, aside from some choppy music!: http://onemorelevel.com/static/games3/burritobison.swf

Fixes #10143.

This is my first contribution to this amazing project, I'm wondering what is the simplest way to create a test for this?

---

Note: the SWF on Newgrounds linked in #10143 issue gives this error screen when testing locally (haven't tried compiling the extension and testing on the web but that may resolve it):

Screenshot 2024-05-06 at 8 36 49 PM

The SWF I linked above works great.

### CODE REVIEW

**Comment by Coder-256** [comment]
I did a quick search for other issues that may be related and it does seem that "Prom Week" from #13250 now gets past the loading screen (but is still unplayable due to graphical glitches).

File: core/src/avm2/globals/avmplus.as [hunk 1]
```diff
@@ -109,6 +109,9 @@ package avmplus {
     }
 
     public function describeType(value: *, flags: uint):XML {
+        if (value == null) {
+            return <type name="null" isDynamic="false" isFinal="true" isStatic="false"/>;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/avmplus.as:113]: From what I can see this should actually be implemented in `describeTypeJSON`. Try comparing the output of `describeType(null,INCLUDE_BASES)` between this branch and FP (with `avmplus.describeType` and `avmplus.INCLUDE_BASES`)

**Comment by Coder-256** [comment]
@Lord-McSweeney Ok, I've moved it to `avmplus.rs`. The new implementation is effectively identical, except it throws an error in a few cases where real FP throws an error.

My version isn't technically true to the original `avmplus`, but it behaves in effectively the same way from the perspective of an external user. Is this better?

File: core/src/avm2/globals/avmplus.rs [hunk 2]
```diff
@@ -80,6 +83,48 @@ bitflags::bitflags! {
...
 
+fn describe_type_json_null<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    flags: DescribeTypeFlags,
+) -> Result<Value<'gc>, Error<'gc>> {
+    // These checks will effectively cause `describeType` to faithfully mirror FP's
+    // implementation. That is to say, it will cause it to throw an "Error #1009: Cannot access
+    // a property or method of a null object reference."
+    if flags.contains(DescribeTypeFlags::USE_ITRAITS)
+        || !flags.contains(DescribeTypeFlags::INCLUDE_BASES | DescribeTypeFlags::INCLUDE_TRAITS)
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/avmplus.rs:94]: Where does this `!flags.contains(DescribeTypeFlags::INCLUDE_BASES | DescribeTypeFlags::INCLUDE_TRAITS)` come from? I haven't been able to get `describeTypeJSON` to return `null` except when `USE_ITRAITS` (the first condition) is set.

Comment (Lord-McSweeney) [reply core/src/avm2/globals/avmplus.rs:94]: By the way you can directly test the output of `describeTypeJSON` with a SWF that uses a function in the `avmplus` package to interface with `describeTypeJSON`, like so:
```
package avmplus
{
    public function describeTypeHack(a:*, b:*) : *
    {
        return describeTypeJSON(a,b);
    }
}
```

File: core/src/avm2/globals/avmplus.rs [hunk 3]
```diff
@@ -80,6 +83,48 @@ bitflags::bitflags! {
...
+
+    traits.set_public_property("bases", ArrayObject::empty(activation)?.into(), activation)?;
+    traits.set_public_property("interfaces", Value::Null, activation)?;
+    traits.set_public_property("variables", Value::Null, activation)?;
+    traits.set_public_property("accessors", Value::Null, activation)?;
+    traits.set_public_property("methods", Value::Null, activation)?;
+    traits.set_public_property("metadata", Value::Null, activation)?;
+    traits.set_public_property("constructor", Value::Null, activation)?;
+
+    object.set_public_property("traits", traits.into(), activation)?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/avmplus.rs:123]: Seems like traits gets set to `null` if `INCLUDE_TRAITS` isn't set

File: core/src/avm2/globals/avmplus.rs [hunk 4]
```diff
@@ -80,6 +83,48 @@ bitflags::bitflags! {
...
+    object.set_public_property("isFinal", true.into(), activation)?;
+    object.set_public_property("isStatic", false.into(), activation)?;
+
+    let traits = activation
+        .avm2()
+        .classes()
+        .object
+        .construct(activation, &[])?;
+
+    traits.set_public_property("bases", ArrayObject::empty(activation)?.into(), activation)?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/avmplus.rs:115]: Seems like `bases` gets set to `null` if `INCLUDE_BASES` isn't set

**Comment by Coder-256** [comment]
@Lord-McSweeney My reasoning was that since `describeTypeJSON` is marked as an internal method, any behavior that ends up identical from the perspective of the end user is acceptable (`describeType` is the only public way to call the method). Is it also a goal to make `describeTypeJSON` behave more closely to FP in case the user uses it via something like reflection? I'm trying to understand why the difference matters if it is an internal method.

**Comment by Coder-256** [comment]
To your points that `traits` is set to null if `INCLUDE_TRAITS` isn't set, and that `bases` is set to null if `INCLUDE_BASES` isn't set, in both cases this triggers a null object error inside `describeType`. That is why I simply chose to short circuit rather than implement these methods more accurately since the fields aren't accessible to code outside the avmplus package if I understand correctly.

**Comment by Coder-256** [comment]
Ok, in this version I now do believe it should fully work correctly even with something like `describeTypeHack`.

**Review by Lord-McSweeney — APPROVED** [review]

**Pull request merged**